### PR TITLE
element-desktop,electron: fix cross build

### PR DESCRIPTION
**Pull request by Mic92**
## Description of changes

#309482 because it fixes an python-dbus cross build, which is quite important

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux (cross from x86_64-linux)
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by Mic92** [comment]
Ok. Found the error for electron finally in my logs:

```
FAILED: clang_x64_v8_arm64/obj/third_party/electron_node/deps/uv/libuv.a
rm -f clang_x64_v8_arm64/obj/third_party/electron_node/deps/uv/libuv.a && "/nix/store/fi0i80hv9wkfqp6xqhdw1sxsmi4fdysr-clang-wrapper-17.0.6/bin/llvm-ar" -T -r -c -s -D clang_x64_v8_arm64/obj/third_party/electron_node/deps/uv/libuv.a @"clang_x64_v8_arm64/obj/third_party/electron_node/deps/uv/libuv.a.rsp"
/bin/sh: /nix/store/fi0i80hv9wkfqp6xqhdw1sxsmi4fdysr-clang-wrapper-17.0.6/bin/llvm-ar: not found
```

**Comment by Mic92** [comment]
Tested with native build of element-desktop and chromium.
pkgsCross.aarch64-multiplatform.element-desktop is still building.

**Comment by Mic92** [comment]
Ok. new issue:

```
"python3" "../../build/toolchain/gcc_solink_wrapper.py" --readelf="$READELF_FOR_BUILD" --nm="$NM_FOR_BUILD"  --sofile="host/libffmpeg.so" --tocfile="host/libffmpeg.so.TOC" --output="host/libffmpeg.so" -- $CXX_FOR_BUILD -shared -Wl,-soname="libffmpeg.so" -Wl,-Bsymbolic -L host/gen/third_party/ffmpeg -fuse-ld=lld -Wl,--fatal-warnings -Wl,--build-id=
sha1 -fPIC -Wl,-z,noexecstack -Wl,-z,relro -Wl,-z,now -Wl,--icf=all -Wl,--color-diagnostics -Wl,--undefined-version -Wl,--no-call-graph-profile-sort -m64 -no-canonical-prefixes -Wl,-O2 -Wl,--gc-sections -Wl,-z,defs -Wl,--as-needed -nostdlib++ -Wl,-rpath=\$ORIGIN -o "host/libffmpeg.so" @"host/libffmpeg.so.rsp"
ld.lld: error: /nix/store/abwx6734x79jib5807nibnl6jlzdrm8b-zlib-aarch64-unknown-linux-gnu-1.3.1/lib/libz.so is incompatible with elf64-x86-64
clang++: error: linker command failed with exit code 1 (use -v to see invocation)
```

**Comment by Mic92** [comment]
Ok. I had forgotten that my chromium patch is still in staging. Now I wait for another rebuild.

**Comment by emilylange** [comment]
Marking as draft because as you said, it does not build/compile yet.

**Comment by Mic92** [comment]
Yeah. I had an issue with building libuv, will get to this bit back laster.

**Comment by Mic92** [comment]
I was able to build  element-desktop, pkgsCross.aarch64-multiplatform.element-desktop, chromium and pkgsCross.aarch64-multiplatform.chromium

**Review by emilylange — APPROVED** [review]
The chromium diff looks good to me.

Can't comment on the element-desktop change though, so maybe wait until one of them approves it as well.

Would be cool to merge this together with #310209 because that also rebuilds all of chromium and electron-source. That way, we could save some hydra resources :)

**Comment by Mic92** [comment]
I don't think it makes much of a difference if these pull requests are merged together because this pull request is targeting staging, where chromium needs to be rebuild anyway independent of master.

**Comment by Mic92** [comment]
cc @SuperSandro2000

**Comment by emilylange** [comment]
> I don't think it makes much of a difference if these pull requests are merged together because this pull request is targeting staging, where chromium needs to be rebuild anyway independent of master.

Whoops, yes, my bad, you are right.

File: pkgs/applications/networking/instant-messengers/element/keytar/default.nix [hunk 1]
```diff
@@ -31,6 +31,17 @@ in stdenv.mkDerivation rec {
...
+  postPatch = lib.optionalString (stdenv.hostPlatform != stdenv.buildPlatform) ''
+    pkg-config() { "''${PKG_CONFIG}" "$@"; }
+    export -f pkg-config
+  '';
+
+  # https://nodejs.org/api/os.html#osarch
+  npmFlagsArray = [ "--arch=${if stdenv.hostPlatform.parsed.cpu.arch == "i686" then "ia32"
+                             else if stdenv.hostPlatform.parsed.cpu.arch == "x86_64" then "x64"
+                             else if stdenv.hostPlatform.parsed.cpu.arch == "aarch64" then "arm64"
+                             else stdenv.hostPlatform.parsed.cpu.arch}" ];
```
Comment (Ma27) [inline pkgs/applications/networking/instant-messengers/element/keytar/default.nix:43]: I have to admit I'm not too familiar with `lib/systems`, but e.g. `nix-instantiate -A pkgsCross.mipsel-linux-gnu.stdenv.hostPlatform.parsed.cpu --eval` doesn't have an arch set.

Comment (Mic92) [reply pkgs/applications/networking/instant-messengers/element/keytar/default.nix]: Looks like this was the wrong attribute to begin with. I am surprised it worked at all.

Comment (Mic92) [reply pkgs/applications/networking/instant-messengers/element/keytar/default.nix]: Make me wonder if it matters at all what we pass in this case.

**Review by Ma27 — APPROVED** [review]
Chromium changes are reviewed (don't know that good enough to comment on that), rest looks good :+1:

**Pull request merged**